tests/jballerina-unit-test/src/test/resources/test-src/query/query_expr_with_errors.bal

```diff
@@ -210,6 +210,50 @@ function throwUnreachableError() returns int|UnreachableError {
     return error UnreachableError("intersection must not be empty", message="GFGF");
 }
 
+function testCatchingErrorAtOnFail() {
+    error? res1 = ();
+    do {
+        _ = from int v in 1 ... 3
+            select check verifyCheck(v);
+    } on fail error err {
+        res1 = err;
+    }
+    assertTrue(res1 is error);
+
+    error? res2 = ();
+    do {
+        do {
+            _ = from int v in 1 ... 3
+                select check verifyCheck(v);
+        } on fail error err {
+            _ = from int v in 1 ... 3
+                select check verifyCheck(v);
+        }
+    } on fail error err {
+        res2 = err;
+    }
+    assertTrue(res2 is error);
+
+    error? res3 = ();
+    do {
+        _ = from int v in 1 ... 3
+            let int intVal = check verifyCheck(v)
+            select 1;
+    } on fail error err {
+        res3 = err;
+    }
+    assertTrue(res3 is error);
+}
+
+function testErrorReturnedFromSelect() {
+    assertTrue(checkErrorAtSelect() is error);
+}
+
+function checkErrorAtSelect() returns error? {
+    _ = from int v in 1 ... 3
+        select check verifyCheck(v);
+}
+
```

Comment:
Doesn't work with `where` clause?
The following crashes.
```ballerina
import ballerina/io;
public function main() returns error? {
    do {
        var ans = from int v in 1 ... 3
            where v == check verifyCheck(1)
            select v;

        io:println(ans);
    } on fail var err {
        io:println(err);
    }
}

public function verifyCheck(int i) returns int|error {
    return error("Verify Check.");
}
```

Shall we also add some tests with nested queries?
Eg:
```ballerina
public function main() returns error? {
    do {
        _ = from int a in (from int v in [1, 2] let int m = check verifyCheck(v) select m)
            select a;
    } on fail var err {
        io:println(err);
    }
}
```

Reply:
Seems to be an issue with how we handle the `where` vs `let` clause during QueryDesugar. The let version works. Will check
```ballerina
function dummyTest() {
    do {
            var ans = from int v in 1 ... 3
                let boolean bool = v == check verifyCheck(1)
                where bool
                select v;

            print(ans);
        } on fail var err {
            print(err);
        }
}

public function verifyCheck(int i) returns int|error {
    return error("Verify Check.");
}

```

Reply:
Created an issue https://github.com/ballerina-platform/ballerina-lang/issues/35885 to handle this separately as the complete pipeline changes when the return type of the lambda function which is used for wrapping where clause changed from `boolean` to `error|boolean`